virtwho/register.py

```diff
@@ -1,5 +1,487 @@
+        """
+        Define subscription-manager command for registering based on the
+        configured parameters.
+        :param second_org: extra org defined only for satellite, use
+            default org as default.
+        :param activ_key: activation_key
+        :return: command string
+        """
+        org = self.default_org
+        if self.register_type == 'satellite':
+            if second_org:
+                org = self.secondary_org
+            if activ_key:
+                ack = self.activation_key
+                cmd = f'subscription-manager register ' \
+                      f'--org="{org}" ' \
+                      f'--activationkey="{ack}"'
+                return cmd
+        cmd = f'subscription-manager register ' \
+              f'--username={self.username} ' \
+              f'--password={self.password} ' \
+              f'--org={org}'
+        return cmd
+
+    def rhsm_conf_set(self, guest=False):
+        """
+        Configure the /etc/rhsm/rhsm.conf before registering.
+        :param guest: configure for guest when guest=True, default is
+            virt-who host.
+        :return: None
+        """
+        ssh = self.ssh(guest=guest)
+        _, _ = ssh.runcmd(f'rm -f {self.rhsm_conf};'
+                          f'cp {self.rhsm_conf_backup} {self.rhsm_conf}')
+        if self.register_type == 'rhsm':
+            cmd = "sed -i -e 's|^hostname.*|hostname = {0}|g' " \
+                  "/etc/rhsm/rhsm.conf".format(self.server)
+            _, _ = ssh.runcmd(cmd)
+
+    def satellite_cert_install(self, guest=False):
+        """
+        Install certificate when registering to satellite.
+        :param guest: install cert for guest when guest=True, default is
+            virt-who host.
+        :return:
+        """
+        if 'satellite' in self.register_type:
+            ssh = self.ssh(guest=guest)
+            host = self.host(guest=guest)
+            cmd = f'rpm -ihv http://{self.server}' \
+                  f'/pub/katello-ca-consumer-latest.noarch.rpm'
+            ret, _ = ssh.runcmd(cmd)
+            if ret != 0:
+                raise FailException(
+                    f'Failed to install satellite certification for {host}')
+
+    def sku_attach(self, pool=None, quantity=None, guest=False):
+        """
+        Attach subscription by Pool ID or --auto.
+        :param pool: Pool ID, attach by --auto when pool=None
+        :param quantity: subscription number to attach, default is auto.
+        :param guest: attach for guest when guest=True, default is
+            virt-who host.
+        :return: tty output.
+        """
+        ssh = self.ssh(guest=guest)
+        host = self.host(guest=guest)
+        cmd = 'subscription-manager attach'
+        if pool:
+            cmd = f'{cmd} --pool={pool}'
+        if quantity:
+            cmd = f'{cmd} --quantity={quantity}'
+        if not pool:
+            cmd = f'{cmd} --auto'
+        for i in range(3):
```

Comment:
remove the loop, we will not use it any more.

Reply:
done, but leave the loop for refresh function. According to my experience the probability of failure is high.